From 77d54a82da83a052a85cbb080ea6931c50e02843 Mon Sep 17 00:00:00 2001
From: Greg McCarthy <greg@gjmccarthy.co.uk>
Date: Thu, 11 Jul 2019 19:04:53 +0100
Subject: [PATCH] Revert "[format] fix include order"

This reverts commit d324e77adc7b31bc9ebd5b4e13513b7fcdcd4bd1.

diff --git a/xbmc/platform/linux/ScreenshotSurfaceRBP.cpp b/xbmc/platform/linux/ScreenshotSurfaceRBP.cpp
index 2c9e0a3b66..def422a19d 100644
--- a/xbmc/platform/linux/ScreenshotSurfaceRBP.cpp
+++ b/xbmc/platform/linux/ScreenshotSurfaceRBP.cpp
@@ -8,9 +8,8 @@
 
 #include "ScreenshotSurfaceRBP.h"
 
-#include "utils/Screenshot.h"
-
 #include "platform/linux/RBP.h"
+#include "utils/Screenshot.h"
 
 void CScreenshotSurfaceRBP::Register()
 {
diff --git a/xbmc/rendering/dx/ScreenshotSurfaceWindows.cpp b/xbmc/rendering/dx/ScreenshotSurfaceWindows.cpp
index cd97e57def..756bc6d098 100644
--- a/xbmc/rendering/dx/ScreenshotSurfaceWindows.cpp
+++ b/xbmc/rendering/dx/ScreenshotSurfaceWindows.cpp
@@ -8,13 +8,13 @@
 
 #include "ScreenshotSurfaceWindows.h"
 
-#include "ServiceBroker.h"
 #include "guilib/GUIComponent.h"
 #include "guilib/GUIWindowManager.h"
 #include "rendering/dx/DeviceResources.h"
+#include "ServiceBroker.h"
 #include "threads/SingleLock.h"
-#include "utils/Screenshot.h"
 #include "utils/log.h"
+#include "utils/Screenshot.h"
 #include "windowing/GraphicContext.h"
 
 #include <wrl/client.h>
diff --git a/xbmc/rendering/gl/ScreenshotSurfaceGL.cpp b/xbmc/rendering/gl/ScreenshotSurfaceGL.cpp
index 1c5589d1b4..be0bd892dd 100644
--- a/xbmc/rendering/gl/ScreenshotSurfaceGL.cpp
+++ b/xbmc/rendering/gl/ScreenshotSurfaceGL.cpp
@@ -8,17 +8,17 @@
 
 #include "ScreenshotSurfaceGL.h"
 
-#include "ServiceBroker.h"
 #include "guilib/GUIComponent.h"
 #include "guilib/GUIWindowManager.h"
+#include "ServiceBroker.h"
 #include "threads/SingleLock.h"
 #include "utils/Screenshot.h"
 #include "windowing/GraphicContext.h"
 
-#include <vector>
-
 #include "system_gl.h"
 
+#include <vector>
+
 void CScreenshotSurfaceGL::Register()
 {
   CScreenShot::Register(CScreenshotSurfaceGL::CreateSurface);
diff --git a/xbmc/rendering/gles/ScreenshotSurfaceGLES.cpp b/xbmc/rendering/gles/ScreenshotSurfaceGLES.cpp
index 2efc744ec7..7d7a9409c2 100644
--- a/xbmc/rendering/gles/ScreenshotSurfaceGLES.cpp
+++ b/xbmc/rendering/gles/ScreenshotSurfaceGLES.cpp
@@ -8,17 +8,17 @@
 
 #include "ScreenshotSurfaceGLES.h"
 
-#include "ServiceBroker.h"
 #include "guilib/GUIComponent.h"
 #include "guilib/GUIWindowManager.h"
+#include "ServiceBroker.h"
 #include "threads/SingleLock.h"
 #include "utils/Screenshot.h"
 #include "windowing/GraphicContext.h"
 
-#include <vector>
-
 #include "system_gl.h"
 
+#include <vector>
+
 void CScreenshotSurfaceGLES::Register()
 {
   CScreenShot::Register(CScreenshotSurfaceGLES::CreateSurface);
diff --git a/xbmc/utils/Screenshot.cpp b/xbmc/utils/Screenshot.cpp
index 638ea64832..c16e001ae4 100644
--- a/xbmc/utils/Screenshot.cpp
+++ b/xbmc/utils/Screenshot.cpp
@@ -8,19 +8,20 @@
 
 #include "Screenshot.h"
 
-#include "ServiceBroker.h"
-#include "URL.h"
-#include "Util.h"
 #include "filesystem/File.h"
 #include "guilib/LocalizeStrings.h"
 #include "pictures/Picture.h"
+#include "ServiceBroker.h"
 #include "settings/SettingPath.h"
 #include "settings/Settings.h"
 #include "settings/SettingsComponent.h"
 #include "settings/windows/GUIControlSettings.h"
+
+#include "URL.h"
+#include "Util.h"
 #include "utils/JobManager.h"
-#include "utils/URIUtils.h"
 #include "utils/log.h"
+#include "utils/URIUtils.h"
 
 using namespace XFILE;
 
diff --git a/xbmc/windowing/rpi/RPIUtils.h b/xbmc/windowing/rpi/RPIUtils.h
index a90c22397c..e4f710920e 100644
--- a/xbmc/windowing/rpi/RPIUtils.h
+++ b/xbmc/windowing/rpi/RPIUtils.h
@@ -8,12 +8,10 @@
 
 #pragma once
 
-#include "windowing/Resolution.h"
-
 #include "platform/linux/RBP.h"
-
 #include <EGL/egl.h>
 #include <bcm_host.h>
+#include "windowing/Resolution.h"
 
 class DllBcmHost;
 class CRPIUtils
diff --git a/xbmc/windowing/rpi/WinSystemRpi.cpp b/xbmc/windowing/rpi/WinSystemRpi.cpp
index 75fbe5602c..bdeb68c8c0 100644
--- a/xbmc/windowing/rpi/WinSystemRpi.cpp
+++ b/xbmc/windowing/rpi/WinSystemRpi.cpp
@@ -8,24 +8,23 @@
 
 #include "WinSystemRpi.h"
 
+#include <string.h>
+#include <float.h>
+
+#include "platform/linux/DllBCM.h"
+#include "platform/linux/RBP.h"
 #include "ServiceBroker.h"
-#include "cores/AudioEngine/AESinkFactory.h"
-#include "cores/AudioEngine/Sinks/AESinkPi.h"
-#include "guilib/DispResource.h"
+#include "windowing/GraphicContext.h"
+#include "windowing/Resolution.h"
 #include "settings/DisplaySettings.h"
 #include "settings/Settings.h"
 #include "settings/SettingsComponent.h"
+#include "guilib/DispResource.h"
 #include "utils/log.h"
-#include "windowing/GraphicContext.h"
-#include "windowing/Resolution.h"
-
-#include "platform/linux/DllBCM.h"
-#include "platform/linux/RBP.h"
+#include "cores/AudioEngine/AESinkFactory.h"
+#include "cores/AudioEngine/Sinks/AESinkPi.h"
 #include "platform/linux/powermanagement/LinuxPowerSyscall.h"
 
-#include <float.h>
-#include <string.h>
-
 #include <EGL/egl.h>
 #include <EGL/eglplatform.h>
 
diff --git a/xbmc/windowing/rpi/WinSystemRpiGLESContext.cpp b/xbmc/windowing/rpi/WinSystemRpiGLESContext.cpp
index 3ec0da452c..06c328ecba 100644
--- a/xbmc/windowing/rpi/WinSystemRpiGLESContext.cpp
+++ b/xbmc/windowing/rpi/WinSystemRpiGLESContext.cpp
@@ -6,24 +6,22 @@
  *  See LICENSES/README.md for more information.
  */
 
-#include "WinSystemRpiGLESContext.h"
-
 #include "Application.h"
-#include "ServiceBroker.h"
 #include "VideoSyncPi.h"
+#include "WinSystemRpiGLESContext.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/GUIWindowManager.h"
+#include "platform/linux/ScreenshotSurfaceRBP.h"
+#include "rendering/gles/ScreenshotSurfaceGLES.h"
+#include "ServiceBroker.h"
+#include "utils/log.h"
 #include "cores/RetroPlayer/process/rbpi/RPProcessInfoPi.h"
 #include "cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGLES.h"
 #include "cores/VideoPlayer/DVDCodecs/DVDFactoryCodec.h"
-#include "cores/VideoPlayer/DVDCodecs/Video/MMALCodec.h"
 #include "cores/VideoPlayer/DVDCodecs/Video/MMALFFmpeg.h"
-#include "cores/VideoPlayer/Process/rbpi/ProcessInfoPi.h"
+#include "cores/VideoPlayer/DVDCodecs/Video/MMALCodec.h"
 #include "cores/VideoPlayer/VideoRenderers/RenderFactory.h"
-#include "guilib/GUIComponent.h"
-#include "guilib/GUIWindowManager.h"
-#include "rendering/gles/ScreenshotSurfaceGLES.h"
-#include "utils/log.h"
-
-#include "platform/linux/ScreenshotSurfaceRBP.h"
+#include "cores/VideoPlayer/Process/rbpi/ProcessInfoPi.h"
 
 using namespace KODI;
 
diff --git a/xbmc/windowing/rpi/WinSystemRpiGLESContext.h b/xbmc/windowing/rpi/WinSystemRpiGLESContext.h
index 0e8ef22a87..74644a560a 100644
--- a/xbmc/windowing/rpi/WinSystemRpiGLESContext.h
+++ b/xbmc/windowing/rpi/WinSystemRpiGLESContext.h
@@ -8,9 +8,9 @@
 
 #pragma once
 
-#include "WinSystemRpi.h"
-#include "rendering/gles/RenderSystemGLES.h"
 #include "utils/EGLUtils.h"
+#include "rendering/gles/RenderSystemGLES.h"
+#include "WinSystemRpi.h"
 
 class CWinSystemRpiGLESContext : public CWinSystemRpi, public CRenderSystemGLES
 {
-- 
2.16.4

