From 58954777a01daf532bf3787546203f5b495688e8 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Sat, 21 Mar 2020 02:21:09 +0100
Subject: [PATCH] Favourites Dialog - fix crash

---
 xbmc/favourites/GUIDialogFavourites.cpp | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/xbmc/favourites/GUIDialogFavourites.cpp b/xbmc/favourites/GUIDialogFavourites.cpp
index 6783a29dfde8..3d35ed228037 100644
--- a/xbmc/favourites/GUIDialogFavourites.cpp
+++ b/xbmc/favourites/GUIDialogFavourites.cpp
@@ -94,10 +94,11 @@ void CGUIDialogFavourites::OnClick(int item)
   if (item < 0 || item >= m_favourites->Size())
     return;
 
-  Close();
-
   CGUIMessage message(GUI_MSG_EXECUTE, 0, GetID());
   message.SetStringParam(m_favouritesService.GetExecutePath(*(*m_favourites)[item], GetID()));
+
+  Close();
+
   CServiceBroker::GetGUI()->GetWindowManager().SendMessage(message);
 }
 
