From df2a32ac79e03d4ba5ad1fb22ac7c4a6a27603a3 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Wed, 3 Apr 2019 18:29:47 +0200
Subject: [PATCH] [python] update WindowXML docs

---
 xbmc/interfaces/legacy/WindowXML.h | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/xbmc/interfaces/legacy/WindowXML.h b/xbmc/interfaces/legacy/WindowXML.h
index 7c2f6422a53b..a7f924818324 100644
--- a/xbmc/interfaces/legacy/WindowXML.h
+++ b/xbmc/interfaces/legacy/WindowXML.h
@@ -44,7 +44,7 @@ namespace XBMCAddon
     ///                             skins path to look in for the xml.
     ///                             (default='Default')
     /// @param defaultRes           [opt] string - default skins resolution.
-    ///                             (default='720p')
+    ///                             (1080i, 720p, ntsc16x9, ntsc, pal16x9 or pal. default='720p')
     /// @param isMedia              [opt] bool - if False, create a regular window.
     ///                             if True, create a mediawindow.
     ///                             (default=False)
@@ -62,7 +62,7 @@ namespace XBMCAddon
     /// **Example:**
     /// ~~~~~~~~~~~~~{.py}
     /// ..
-    /// win = xbmcgui.WindowXML('script-Lyrics-main.xml', xbmcaddon.Addon().getAddonInfo('path').decode('utf-8'), 'default', '1080p', False)
+    /// win = xbmcgui.WindowXML('script-Lyrics-main.xml', xbmcaddon.Addon().getAddonInfo('path').decode('utf-8'), 'default', '1080i', False)
     /// win.doModal()
     /// del win
     /// ..
@@ -483,7 +483,7 @@ namespace XBMCAddon
     ///                                 skins path to look in for the xml.
     ///                                 (default='Default')
     /// @param defaultRes               [opt] string - default skins resolution.
-    ///                                 (default='720p')
+    ///                                 (1080i, 720p, ntsc16x9, ntsc, pal16x9 or pal. default='720p')
     /// @throws Exception               if more then 200 windows are created.
     ///
     /// @note Skin folder structure is e.g. **resources/skins/Default/720p**
@@ -494,7 +494,7 @@ namespace XBMCAddon
     /// **Example:**
     /// ~~~~~~~~~~~~~{.py}
     /// ..
-    /// dialog = xbmcgui.WindowXMLDialog('script-Lyrics-main.xml', xbmcaddon.Addon().getAddonInfo('path').decode('utf-8'), 'default', '1080p')
+    /// dialog = xbmcgui.WindowXMLDialog('script-Lyrics-main.xml', xbmcaddon.Addon().getAddonInfo('path').decode('utf-8'), 'default', '1080i')
     /// dialog.doModal()
     /// del dialog
     /// ..
