From afcf96639d3e7f6badf8ed6f32385aaa140af49f Mon Sep 17 00:00:00 2001
From: James Hutchinson <jahutchinson99@googlemail.com>
Date: Mon, 16 Mar 2020 16:24:11 +0000
Subject: [PATCH] [PVR] add watched count to grouped recordings list

---
 xbmc/pvr/filesystem/PVRGUIDirectory.cpp | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/xbmc/pvr/filesystem/PVRGUIDirectory.cpp b/xbmc/pvr/filesystem/PVRGUIDirectory.cpp
index f5e52589900d..f2d2ec95cfcf 100644
--- a/xbmc/pvr/filesystem/PVRGUIDirectory.cpp
+++ b/xbmc/pvr/filesystem/PVRGUIDirectory.cpp
@@ -202,6 +202,8 @@ void GetSubDirectories(const CPVRRecordingsPath& recParentPath,
       item->SetLabel(strCurrent);
       item->SetLabelPreformatted(true);
       item->m_dateTime = recording->RecordingTimeAsLocalTime();
+      item->SetProperty("totalepisodes", 0);
+      item->SetProperty("watchedepisodes", 0);
       item->SetProperty("unwatchedepisodes", 0);
 
       // Assume all folders are watched, we'll change the overlay later
@@ -215,11 +217,19 @@ void GetSubDirectories(const CPVRRecordingsPath& recParentPath,
         item->m_dateTime = recording->RecordingTimeAsLocalTime();
     }
 
+    item->IncrementProperty("totalepisodes", 1);
     if (recording->GetPlayCount() == 0)
     {
       unwatchedFolders.insert(item);
       item->IncrementProperty("unwatchedepisodes", 1);
     }
+    else
+    {
+      item->IncrementProperty("watchedepisodes", 1);
+    }
+    item->SetLabel2(StringUtils::Format("%s / %s",
+        item->GetProperty("watchedepisodes").asString().c_str(),
+        item->GetProperty("totalepisodes").asString().c_str()));
   }
 
   // Change the watched overlay to unwatched for folders containing unwatched entries
