From a7debcc3733efe848d3a178cbcd3adaa6646fc9b Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Sun, 5 Apr 2020 17:11:38 +0200
Subject: [PATCH] Don't send UNFOCUS_ALL after touch action

---
 xbmc/input/InputManager.cpp | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/xbmc/input/InputManager.cpp b/xbmc/input/InputManager.cpp
index 24dd4d7c731c..594c187a9ae1 100644
--- a/xbmc/input/InputManager.cpp
+++ b/xbmc/input/InputManager.cpp
@@ -418,12 +418,6 @@ bool CInputManager::OnEvent(XBMC_Event& newEvent)
         CApplicationMessenger::GetInstance().PostMsg(TMSG_GUI_ACTION, WINDOW_INVALID, -1, static_cast<void*>(new CAction(actionId)));
     }
 
-    // Post an unfocus message for touch device after the action.
-    if (newEvent.touch.action == ACTION_GESTURE_END || newEvent.touch.action == ACTION_TOUCH_TAP)
-    {
-      CGUIMessage msg(GUI_MSG_UNFOCUS_ALL, 0, 0, 0, 0);
-      CApplicationMessenger::GetInstance().SendGUIMessage(msg);
-    }
     break;
   } //case
   case XBMC_BUTTON:
