From 4f6050d6d99b104e664543e32ae21b87e4d0e040 Mon Sep 17 00:00:00 2001
From: Kai Sommerfeld <kai.sommerfeld@gmx.com>
Date: Thu, 8 Aug 2019 23:05:21 +0200
Subject: [PATCH] [dialogs] Fix missing cancel button in progress dialogs.

---
 xbmc/dialogs/GUIDialogProgress.cpp | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/xbmc/dialogs/GUIDialogProgress.cpp b/xbmc/dialogs/GUIDialogProgress.cpp
index b0fac82b4b10..a64533b9b110 100644
--- a/xbmc/dialogs/GUIDialogProgress.cpp
+++ b/xbmc/dialogs/GUIDialogProgress.cpp
@@ -230,19 +230,21 @@ void CGUIDialogProgress::UpdateControls()
   else
     SET_CONTROL_HIDDEN(CONTROL_PROGRESS_BAR);
 
-  // special handling for choice 0 mapped to cancel button
-  if (bShowCancel)
-    SET_CONTROL_VISIBLE(CONTROL_CHOICES_START);
-  else
-    SET_CONTROL_HIDDEN(CONTROL_CHOICES_START);
-
+  bool bAllHidden = true;
   for (int i = 0; i < DIALOG_MAX_CHOICES; ++i)
   {
     if (choices[i])
+    {
+      bAllHidden = false;
       SET_CONTROL_VISIBLE(CONTROL_CHOICES_START + i);
+    }
     else
       SET_CONTROL_HIDDEN(CONTROL_CHOICES_START + i);
   }
+
+  // special handling for choice 0 mapped to cancel button
+  if (bShowCancel && bAllHidden)
+    SET_CONTROL_VISIBLE(CONTROL_CHOICES_START);
 }
 
 void CGUIDialogProgress::Process(unsigned int currentTime, CDirtyRegionList &dirtyregions)
@@ -257,15 +259,21 @@ void CGUIDialogProgress::OnInitWindow()
 {
   UpdateControls();
 
+  bool bNoFocus = true;
   for (int i = 0; i < DIALOG_MAX_CHOICES; ++i)
   {
     if (m_supportedChoices[i])
     {
+      bNoFocus = false;
       SET_CONTROL_FOCUS(CONTROL_CHOICES_START + i, 0);
       break;
     }
   }
 
+  // special handling for choice 0 mapped to cancel button
+  if (m_bCanCancel && bNoFocus)
+    SET_CONTROL_FOCUS(CONTROL_CHOICES_START,0 );
+
   CGUIDialogBoxBase::OnInitWindow();
 }
 
