From 4f05fabf8feb299e1f0780d4e9a735ab3202dcc0 Mon Sep 17 00:00:00 2001
From: phunkyfish <phunkyfish@gmail.com>
Date: Sun, 26 Apr 2020 12:12:39 +0100
Subject: [PATCH] [pvr] get previous epg tag instead of previous of previous

---
 xbmc/pvr/epg/EpgTagsCache.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/pvr/epg/EpgTagsCache.cpp b/xbmc/pvr/epg/EpgTagsCache.cpp
index 34f628762643..78e0e45ec7c2 100644
--- a/xbmc/pvr/epg/EpgTagsCache.cpp
+++ b/xbmc/pvr/epg/EpgTagsCache.cpp
@@ -127,7 +127,7 @@ void CPVREpgTagsCache::RefreshLastEndedTag(const CDateTime& activeTime)
 {
   if (m_database)
   {
-    m_lastEndedTag = m_database->GetEpgTagByMaxEndTime(m_iEpgID, activeTime - ONE_SECOND);
+    m_lastEndedTag = m_database->GetEpgTagByMaxEndTime(m_iEpgID, activeTime);
     if (m_lastEndedTag)
       m_lastEndedTag->SetChannelData(m_channelData);
   }
