From 76b7ffb66949c7d31f7a812e4cd47df64a5ba86d Mon Sep 17 00:00:00 2001
From: howie-f <rftc@gmx.de>
Date: Sun, 12 Apr 2020 23:10:56 +0200
Subject: [PATCH] [clang-format] disable for all arrays in LangCodeExpander.cpp

---
 xbmc/utils/LangCodeExpander.cpp | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/xbmc/utils/LangCodeExpander.cpp b/xbmc/utils/LangCodeExpander.cpp
index 014e9be820c3..f0b21ed32fe6 100644
--- a/xbmc/utils/LangCodeExpander.cpp
+++ b/xbmc/utils/LangCodeExpander.cpp
@@ -547,6 +547,7 @@ std::string CLangCodeExpander::ConvertToISO6392T(const std::string& lang)
   return lang;
 }
 
+// clang-format off
 const std::array<struct LCENTRY, 185> g_iso639_1 = {{
     {MAKECODE('\0', '\0', 'a', 'a'), "Afar"},
     {MAKECODE('\0', '\0', 'a', 'b'), "Abkhazian"},
@@ -734,7 +735,9 @@ const std::array<struct LCENTRY, 185> g_iso639_1 = {{
     {MAKECODE('\0', '\0', 'z', 'h'), "Chinese"},
     {MAKECODE('\0', '\0', 'z', 'u'), "Zulu"},
 }};
+// clang-format on
 
+// clang-format off
 const std::array<struct LCENTRY, 539> g_iso639_2 = {{
     {MAKECODE('\0', 'a', 'b', 'k'), "Abkhaz"},
     {MAKECODE('\0', 'a', 'b', 'k'), "Abkhazian"},
@@ -1279,8 +1282,9 @@ const std::array<struct LCENTRY, 539> g_iso639_2 = {{
     {MAKECODE('\0', 'z', 'u', 'l'), "Zulu"},
     {MAKECODE('\0', 'z', 'u', 'n'), "Zuni"},
 }};
+// clang-format on
 
-
+// clang-format off
 const std::array<ISO639, 189> LanguageCodes = {{
     {"aa", "aar", NULL, NULL},
     {"ab", "abk", NULL, NULL},
@@ -1474,6 +1478,7 @@ const std::array<ISO639, 189> LanguageCodes = {{
      NULL}, // Kodi intern mapping for missing "Miscellaneous languages" iso639-1 code
     {"zz", "mul", NULL, NULL} // Kodi intern mapping for missing "Multiple languages" iso639-1 code
 }};
+// clang-format on
 
 // Based on ISO 3166
 // clang-format off
