From 7afc2deadb1e2762d98cb07999b667cbea573571 Mon Sep 17 00:00:00 2001
From: mediaminister <mediaminister@gmail.com>
Date: Wed, 15 Apr 2020 21:42:47 +0200
Subject: [PATCH] Support multi-line text in dialogs

---
 xbmc/interfaces/legacy/Dialog.cpp | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/xbmc/interfaces/legacy/Dialog.cpp b/xbmc/interfaces/legacy/Dialog.cpp
index af5f811d19a0..608f1e4032ac 100644
--- a/xbmc/interfaces/legacy/Dialog.cpp
+++ b/xbmc/interfaces/legacy/Dialog.cpp
@@ -53,11 +53,10 @@ namespace XBMCAddon
       if (pDialog == NULL)
         throw WindowException("Error: Window is NULL, this is not possible :-)");
 
-      // get lines, last 4 lines are optional.
       if (!heading.empty())
         pDialog->SetHeading(CVariant{heading});
       if (!message.empty())
-        pDialog->SetLine(0, CVariant{message});
+        pDialog->SetText(CVariant{message});
 
       if (!nolabel.empty())
         pDialog->SetChoice(0, CVariant{nolabel});
@@ -449,7 +448,7 @@ namespace XBMCAddon
       pDialog->SetHeading(CVariant{heading});
 
       if (!message.empty())
-        pDialog->SetLine(0, CVariant{message});
+        pDialog->SetText(CVariant{message});
 
       pDialog->Open();
     }
@@ -473,7 +472,7 @@ namespace XBMCAddon
       }
 
       if (!message.empty())
-        pDialog->SetLine(0, CVariant{message});
+        pDialog->SetText(CVariant{message});
     }
 
     void DialogProgress::close()
