From 2822073039566d30bac659712c9667b4bcc36811 Mon Sep 17 00:00:00 2001
From: mglae <mglmail@arcor.de>
Date: Thu, 11 Jun 2020 19:05:23 +0200
Subject: [PATCH] EDL comskip: support commercial at 0:00

---
 xbmc/cores/VideoPlayer/Edl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/cores/VideoPlayer/Edl.cpp b/xbmc/cores/VideoPlayer/Edl.cpp
index d3c533c3d95d..51c14ed5ac89 100644
--- a/xbmc/cores/VideoPlayer/Edl.cpp
+++ b/xbmc/cores/VideoPlayer/Edl.cpp
@@ -40,7 +40,7 @@ void CEdl::Clear()
   m_vecCuts.clear();
   m_vecSceneMarkers.clear();
   m_iTotalCutTime = 0;
-  m_lastCutTime = 0;
+  m_lastCutTime = -1;
 }
 
 bool CEdl::ReadEditDecisionLists(const CFileItem& fileItem, const float fFramesPerSecond)
