From ffdab825561ff96799637a73bdcbce955ea658d9 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Sun, 3 Mar 2019 22:30:07 +0100
Subject: [PATCH 1/2] [python] file.seek() - make whence optional

---
 xbmc/interfaces/legacy/File.h | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/xbmc/interfaces/legacy/File.h b/xbmc/interfaces/legacy/File.h
index 9eb18a05c2a4..a947c4e92e49 100644
--- a/xbmc/interfaces/legacy/File.h
+++ b/xbmc/interfaces/legacy/File.h
@@ -182,11 +182,12 @@ namespace XBMCAddon
       /// Seek to position in file.
       ///
       /// @param seekBytes          position in the file
-      /// @param iWhence            where in a file to seek from[0 beginning,
+      /// @param iWhence            [opt] where in a file to seek from[0 beginning,
       ///                           1 current , 2 end position]
       ///
       ///
       ///-----------------------------------------------------------------------
+      /// @python_v19 Function changed. param **iWhence** is now optional.
       ///
       /// **Example:**
       /// ~~~~~~~~~~~~~{.py}
@@ -199,7 +200,7 @@ namespace XBMCAddon
       ///
       seek(...);
 #else
-      inline long long seek(long long seekBytes, int iWhence) { DelayedCallGuard dg(languageHook); return file->Seek(seekBytes,iWhence); }
+      inline long long seek(long long seekBytes, int iWhence = SEEK_SET) { DelayedCallGuard dg(languageHook); return file->Seek(seekBytes,iWhence); }
 #endif
 
 #ifdef DOXYGEN_SHOULD_USE_THIS

From d2d273c518a20e188f1d64252c7e50ed547596a8 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Sun, 3 Mar 2019 22:32:03 +0100
Subject: [PATCH 2/2] [python-docs] xbmcvfs - fix file.readBytes() example

---
 xbmc/interfaces/legacy/File.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/interfaces/legacy/File.h b/xbmc/interfaces/legacy/File.h
index a947c4e92e49..5e2a53618da1 100644
--- a/xbmc/interfaces/legacy/File.h
+++ b/xbmc/interfaces/legacy/File.h
@@ -111,7 +111,7 @@ namespace XBMCAddon
       /// ~~~~~~~~~~~~~{.py}
       /// ..
       /// f = xbmcvfs.File(file)
-      /// b = f.read()
+      /// b = f.readBytes()
       /// f.close()
       /// ..
       /// ~~~~~~~~~~~~~
