From 9fd63506d82809223d6af6f9eda73faabd47d51a Mon Sep 17 00:00:00 2001
From: Arne Morten Kvarving <spiff@kodi.tv>
Date: Mon, 5 Aug 2019 10:32:02 +0200
Subject: [PATCH 1/2] fixed: use first protocol from add-on in add network
 dialog

---
 xbmc/network/GUIDialogNetworkSetup.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/xbmc/network/GUIDialogNetworkSetup.cpp b/xbmc/network/GUIDialogNetworkSetup.cpp
index 4aaad76539c6..0ccac9f33a28 100644
--- a/xbmc/network/GUIDialogNetworkSetup.cpp
+++ b/xbmc/network/GUIDialogNetworkSetup.cpp
@@ -445,10 +445,14 @@ void CGUIDialogNetworkSetup::UpdateAvailableProtocols()
     {
       const auto& info = addon->GetProtocolInfo();
       if (!addon->GetProtocolInfo().type.empty())
+      {
+        // only use first protocol
+        auto prots = StringUtils::Split(info.type, "|");
         m_protocols.emplace_back(Protocol{ info.supportPath, info.supportUsername,
           info.supportPassword, info.supportPort,
           info.supportBrowsing, info.defaultPort,
-          info.type, info.label });
+          prots.front(), info.label });
+      }
     }
   }
   // internals

From 0ca7db8bf6d6d4e3229bc9ace90fe301a1b35abf Mon Sep 17 00:00:00 2001
From: Arne Morten Kvarving <spiff@kodi.tv>
Date: Mon, 5 Aug 2019 10:59:01 +0200
Subject: [PATCH 2/2] fixed: since there may be multiple protocols, we have to
 split

---
 xbmc/Util.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/xbmc/Util.cpp b/xbmc/Util.cpp
index 62196a4ec844..a5f9c26f7d5d 100644
--- a/xbmc/Util.cpp
+++ b/xbmc/Util.cpp
@@ -1512,7 +1512,9 @@ bool CUtil::SupportsWriteFileOperations(const std::string& strPath)
     for (const auto& addon : CServiceBroker::GetVFSAddonCache().GetAddonInstances())
     {
       const auto& info = addon->GetProtocolInfo();
-      if (info.type == url.GetProtocol() && info.supportWrite)
+      auto prots = StringUtils::Split(info.type, "|");
+      if (info.supportWrite &&
+          std::find(prots.begin(), prots.end(), url.GetProtocol()) != prots.end())
         return true;
     }
   }
