From c371d725b74f00e78fff497243b6ac09e076b2cb Mon Sep 17 00:00:00 2001
From: Arthur Liberman <arthur_liberman@hotmail.com>
Date: Sat, 9 Nov 2019 14:39:33 +0200
Subject: [PATCH] fixed: redact url in log

---
 xbmc/filesystem/MultiPathDirectory.cpp | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/xbmc/filesystem/MultiPathDirectory.cpp b/xbmc/filesystem/MultiPathDirectory.cpp
index b313a69e6349..789a0b2beb45 100644
--- a/xbmc/filesystem/MultiPathDirectory.cpp
+++ b/xbmc/filesystem/MultiPathDirectory.cpp
@@ -74,12 +74,12 @@ bool CMultiPathDirectory::GetDirectory(const CURL& url, CFileItemList &items)
     }
 
     CFileItemList tempItems;
-    CLog::Log(LOGDEBUG,"Getting Directory (%s)", vecPaths[i].c_str());
+    CLog::Log(LOGDEBUG,"Getting Directory (%s)", CURL::GetRedacted(vecPaths[i]).c_str());
     if (CDirectory::GetDirectory(vecPaths[i], tempItems, m_strFileMask, m_flags))
       items.Append(tempItems);
     else
     {
-      CLog::Log(LOGERROR,"Error Getting Directory (%s)", vecPaths[i].c_str());
+      CLog::Log(LOGERROR,"Error Getting Directory (%s)", CURL::GetRedacted(vecPaths[i]).c_str());
       iFailures++;
     }
 
@@ -112,7 +112,7 @@ bool CMultiPathDirectory::Exists(const CURL& url)
 
   for (unsigned int i = 0; i < vecPaths.size(); ++i)
   {
-    CLog::Log(LOGDEBUG,"Testing Existence (%s)", vecPaths[i].c_str());
+    CLog::Log(LOGDEBUG,"Testing Existence (%s)", CURL::GetRedacted(vecPaths[i]).c_str());
     if (CDirectory::Exists(vecPaths[i]))
       return true;
   }
@@ -255,7 +255,7 @@ void CMultiPathDirectory::MergeItems(CFileItemList &items)
 
     std::vector<int> stack;
     stack.push_back(i);
-    CLog::Log(LOGDEBUG,"Testing path: [%03i] %s", i, pItem1->GetPath().c_str());
+    CLog::Log(LOGDEBUG,"Testing path: [%03i] %s", i, CURL::GetRedacted(pItem1->GetPath()).c_str());
 
     int j = i + 1;
     do
@@ -269,7 +269,7 @@ void CMultiPathDirectory::MergeItems(CFileItemList &items)
       if (!pItem2->IsFileFolder())
       {
         stack.push_back(j);
-        CLog::Log(LOGDEBUG,"  Adding path: [%03i] %s", j, pItem2->GetPath().c_str());
+        CLog::Log(LOGDEBUG,"  Adding path: [%03i] %s", j, CURL::GetRedacted(pItem2->GetPath()).c_str());
       }
       j++;
     }
@@ -283,7 +283,7 @@ void CMultiPathDirectory::MergeItems(CFileItemList &items)
       for (unsigned int k = stack.size() - 1; k > 0; --k)
         items.Remove(stack[k]);
       pItem1->SetPath(newPath);
-      CLog::Log(LOGDEBUG,"  New path: %s", pItem1->GetPath().c_str());
+      CLog::Log(LOGDEBUG,"  New path: %s", CURL::GetRedacted(pItem1->GetPath()).c_str());
     }
 
     i++;
