From 388bfe99a4687dea336800f256a7994a3e4b228f Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Sat, 7 Sep 2019 22:14:56 +0200
Subject: [PATCH] [EGL Render] Fix accidentially changed egl config

---
 xbmc/utils/EGLUtils.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xbmc/utils/EGLUtils.cpp b/xbmc/utils/EGLUtils.cpp
index 1fdc9dedaf5f..8f300dbaedaf 100644
--- a/xbmc/utils/EGLUtils.cpp
+++ b/xbmc/utils/EGLUtils.cpp
@@ -310,7 +310,7 @@ bool CEGLContextUtils::ChooseConfig(EGLint renderableType, EGLint visualId, bool
     throw std::logic_error("Choosing an EGLConfig requires an EGL display");
   }
 
-  EGLint surfaceType = EGL_WINDOW_BIT | EGL_PBUFFER_BIT;
+  EGLint surfaceType = EGL_WINDOW_BIT;
   // for the non-trivial dirty region modes, we need the EGL buffer to be preserved across updates
   int guiAlgorithmDirtyRegions = CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_guiAlgorithmDirtyRegions;
   if (guiAlgorithmDirtyRegions == DIRTYREGION_SOLVER_COST_REDUCTION ||
@@ -321,7 +321,7 @@ bool CEGLContextUtils::ChooseConfig(EGLint renderableType, EGLint visualId, bool
   attribs.Add({{EGL_RED_SIZE, 8},
                {EGL_GREEN_SIZE, 8},
                {EGL_BLUE_SIZE, 8},
-               {EGL_ALPHA_SIZE, 8},
+               {EGL_ALPHA_SIZE, 2},
                {EGL_DEPTH_SIZE, 16},
                {EGL_STENCIL_SIZE, 0},
                {EGL_SAMPLE_BUFFERS, 0},
