From 70e3d4eedf590ed70194d82255c69695cfb5d5a2 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Mon, 25 Mar 2019 16:47:51 +0100
Subject: [PATCH] addon settings - inherit settings level

---
 xbmc/addons/settings/GUIDialogAddonSettings.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/xbmc/addons/settings/GUIDialogAddonSettings.cpp b/xbmc/addons/settings/GUIDialogAddonSettings.cpp
index 4c642d00e8e8..0050a7e98ccd 100644
--- a/xbmc/addons/settings/GUIDialogAddonSettings.cpp
+++ b/xbmc/addons/settings/GUIDialogAddonSettings.cpp
@@ -19,6 +19,7 @@
 #include "settings/lib/SettingsManager.h"
 #include "utils/StringUtils.h"
 #include "utils/Variant.h"
+#include "view/ViewStateSettings.h"
 
 using namespace KODI::MESSAGING;
 
@@ -171,7 +172,7 @@ std::string CGUIDialogAddonSettings::GetSettingsLabel(std::shared_ptr<ISetting>
 
 int CGUIDialogAddonSettings::GetSettingLevel() const
 {
-  return static_cast<int>(SettingLevel::Standard);
+  return static_cast<int>(CViewStateSettings::GetInstance().GetSettingLevel());
 }
 
 std::shared_ptr<CSettingSection> CGUIDialogAddonSettings::GetSection()
