From 18aa94d444f85a8fd2331eff4cb42967f087d3f0 Mon Sep 17 00:00:00 2001
From: DaveTBlake <oak99sky@yahoo.co.uk>
Date: Thu, 26 Sep 2019 16:29:03 +0100
Subject: [PATCH] Clear "libraryartfilled" property when art map is cleared

---
 xbmc/guilib/GUIListItem.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/xbmc/guilib/GUIListItem.cpp b/xbmc/guilib/GUIListItem.cpp
index 379fd9de5a81..6e0ab9f23945 100644
--- a/xbmc/guilib/GUIListItem.cpp
+++ b/xbmc/guilib/GUIListItem.cpp
@@ -120,6 +120,7 @@ void CGUIListItem::ClearArt()
 {
   m_art.clear();
   m_artFallbacks.clear();
+  SetProperty("libraryartfilled", false);
 }
 
 void CGUIListItem::AppendArt(const ArtMap &art, const std::string &prefix)
