From 0eb5728927223378b7f92bf5631832a0e6a10d08 Mon Sep 17 00:00:00 2001
From: djp952 <djp952@gmail.com>
Date: Tue, 26 Nov 2019 23:25:32 -0500
Subject: [PATCH] [Visualization] Prevent NULL/zero-length packets from being
 processed

---
 xbmc/guilib/GUIVisualisationControl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/guilib/GUIVisualisationControl.cpp b/xbmc/guilib/GUIVisualisationControl.cpp
index c63751b06610..3a4bddf94623 100644
--- a/xbmc/guilib/GUIVisualisationControl.cpp
+++ b/xbmc/guilib/GUIVisualisationControl.cpp
@@ -242,7 +242,7 @@ void CGUIVisualisationControl::OnInitialize(int channels, int samplesPerSec, int
 
 void CGUIVisualisationControl::OnAudioData(const float* audioData, unsigned int audioDataLength)
 {
-  if (!m_instance || !m_alreadyStarted)
+  if (!m_instance || !m_alreadyStarted || !audioData || audioDataLength == 0)
     return;
 
   // Save our audio data in the buffers
