From e724a7b06f88df040904062d47d2fdba5d1766bd Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Mon, 8 Apr 2019 09:47:18 +0200
Subject: [PATCH] [Android] Prevent modeswitch when native_window is invalid

---
 xbmc/windowing/android/WinSystemAndroid.cpp | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/xbmc/windowing/android/WinSystemAndroid.cpp b/xbmc/windowing/android/WinSystemAndroid.cpp
index df4bf99161b3..3790d3e0a0ee 100644
--- a/xbmc/windowing/android/WinSystemAndroid.cpp
+++ b/xbmc/windowing/android/WinSystemAndroid.cpp
@@ -127,11 +127,21 @@ bool CWinSystemAndroid::CreateNewWindow(const std::string& name,
     return true;
   }
 
+  if (m_dispResetState != RESET_NOTWAITING)
+  {
+    CLog::Log(LOGINFO, "CWinSystemAndroid::CreateNewWindow: cannot create window while resetting");
+    return false;
+  }
+
   m_stereo_mode = stereo_mode;
   m_bFullScreen = fullScreen;
 
   m_nativeWindow = CXBMCApp::GetNativeWindow(2000);
-
+  if (!m_nativeWindow)
+  {
+    CLog::Log(LOGERROR, "CWinSystemAndroid::CreateNewWindow: failed");
+    return false;
+  }
   m_android->SetNativeResolution(res);
 
   return true;
