From 93c57211dfd1b7f8052f801389a606bc53e1bb04 Mon Sep 17 00:00:00 2001
From: Rechi <Rechi@users.noreply.github.com>
Date: Mon, 9 Sep 2019 08:37:15 +0200
Subject: [PATCH] [depends] bump freetype2 to 2.10.1

---
 tools/depends/target/freetype2/Makefile | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/tools/depends/target/freetype2/Makefile b/tools/depends/target/freetype2/Makefile
index 51e46442b0dd..a2c82b757b20 100644
--- a/tools/depends/target/freetype2/Makefile
+++ b/tools/depends/target/freetype2/Makefile
@@ -3,9 +3,9 @@ DEPS= ../../Makefile.include Makefile
 
 # lib name, version
 LIBNAME=freetype
-VERSION=2.8
+VERSION=2.10.1
 SOURCE=$(LIBNAME)-$(VERSION)
-ARCHIVE=$(SOURCE).tar.bz2
+ARCHIVE=$(SOURCE).tar.xz
 
 # configuration settings
 # force using internal libtool
@@ -38,4 +38,3 @@ clean:
 
 distclean::
 	rm -rf $(PLATFORM) .installed-$(PLATFORM)
-
