From c14acefbc1f52b49d5c1a4ff92c00b524d5be7a6 Mon Sep 17 00:00:00 2001
From: DaveTBlake <oak99sky@yahoo.co.uk>
Date: Sat, 25 Jan 2020 18:35:15 +0000
Subject: [PATCH] Fix MUSICPLAYER_DISC_TITLE so with other infolabels

---
 xbmc/guilib/guiinfo/GUIInfoLabels.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/guilib/guiinfo/GUIInfoLabels.h b/xbmc/guilib/guiinfo/GUIInfoLabels.h
index c4ee80845c06..9dbcff5a5ab8 100644
--- a/xbmc/guilib/guiinfo/GUIInfoLabels.h
+++ b/xbmc/guilib/guiinfo/GUIInfoLabels.h
@@ -185,6 +185,7 @@
 #define MUSICPLAYER_CONTRIBUTORS    229
 #define MUSICPLAYER_CONTRIBUTOR_AND_ROLE 230
 #define MUSICPLAYER_DBID            231
+#define MUSICPLAYER_DISC_TITLE      232
 
 // Range of musicplayer infolabels that work with offset and position
 #define MUSICPLAYER_OFFSET_POSITION_FIRST MUSICPLAYER_TITLE
@@ -200,7 +201,6 @@
 #define MUSICPLAYER_EXISTS          242
 #define MUSICPLAYER_PLAYLISTPLAYING 243
 #define MUSICPLAYER_CONTENT         244
-#define MUSICPLAYER_DISC_TITLE      245
 
 #define VIDEOPLAYER_AUDIO_BITRATE     248
 #define VIDEOPLAYER_VIDEO_BITRATE     249
