From 9d54869b1e02f64e6adefac87ba2f61cf8b37fac Mon Sep 17 00:00:00 2001
From: arnova <nospam@void.org>
Date: Mon, 30 Sep 2019 10:13:27 +0200
Subject: [PATCH] fixed: Double redact

---
 xbmc/video/VideoDatabase.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/video/VideoDatabase.cpp b/xbmc/video/VideoDatabase.cpp
index de4ab8a22fed..7cb3400391fe 100644
--- a/xbmc/video/VideoDatabase.cpp
+++ b/xbmc/video/VideoDatabase.cpp
@@ -5907,7 +5907,7 @@ void CVideoDatabase::EraseVideoSettings(const CFileItem &item)
   {
     std::string sql = PrepareSQL("DELETE FROM settings WHERE idFile=%i", idFile);
 
-    CLog::Log(LOGINFO, "Deleting settings information for files %s", CURL::GetRedacted(CURL::GetRedacted(item.GetPath()).c_str()));
+    CLog::Log(LOGINFO, "Deleting settings information for files %s", CURL::GetRedacted(item.GetPath()).c_str());
     m_pDS->exec(sql);
   }
   catch (...)
