From 30b5b8dbe6a66babe584d58e74a156d04b157d4c Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Mon, 30 Mar 2020 00:19:14 +0200
Subject: [PATCH] fix VideoPlayer.UniqueID() infolabel

---
 xbmc/GUIInfoManager.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/xbmc/GUIInfoManager.cpp b/xbmc/GUIInfoManager.cpp
index 05198d09ee47..6dbb72354439 100644
--- a/xbmc/GUIInfoManager.cpp
+++ b/xbmc/GUIInfoManager.cpp
@@ -9610,6 +9610,10 @@ int CGUIInfoManager::TranslateSingleString(const std::string &strCondition, bool
       {
         return AddMultiInfo(CGUIInfo(VIDEOPLAYER_CONTENT, prop.param(), 0));
       }
+      if (prop.name == "uniqueid" && prop.num_params())
+      {
+        return AddMultiInfo(CGUIInfo(VIDEOPLAYER_UNIQUEID, prop.param(), 0));
+      }
       return TranslateVideoPlayerString(prop.name);
     }
     else if (cat.name == "retroplayer")
