From f5b5419e9bf281ceaf698c632316305930eb6130 Mon Sep 17 00:00:00 2001
From: enen92 <92enen@gmail.com>
Date: Fri, 17 May 2019 19:54:42 +0100
Subject: [PATCH] [PlaylistPlayer] Resolve plugin paths to fill the item
 infotag

---
 xbmc/PlayListPlayer.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/xbmc/PlayListPlayer.cpp b/xbmc/PlayListPlayer.cpp
index abb080c3a56c..a556ac2c19e7 100644
--- a/xbmc/PlayListPlayer.cpp
+++ b/xbmc/PlayListPlayer.cpp
@@ -25,8 +25,10 @@
 #include "interfaces/AnnouncementManager.h"
 #include "input/Key.h"
 #include "URL.h"
+#include "utils/URIUtils.h"
 #include "messaging/ApplicationMessenger.h"
 #include "filesystem/VideoDatabaseFile.h"
+#include "filesystem/PluginDirectory.h"
 #include "messaging/helpers/DialogOKHelper.h"
 #include "ServiceBroker.h"
 
@@ -892,6 +894,12 @@ void PLAYLIST::CPlayListPlayer::OnApplicationMessage(KODI::MESSAGING::ThreadMess
         if (list->Size() == 1 && !(*list)[0]->IsPlayList())
         {
           CFileItemPtr item = (*list)[0];
+          // if the item is a plugin we need to resolve the URL to ensure the infotags are filled.
+          // resolve only for a maximum of 5 times to avoid deadlocks (plugin:// paths can resolve to plugin:// paths)
+          for (int i = 0; URIUtils::IsPlugin(item->GetDynPath()) && i < 5; ++i)
+          {
+            XFILE::CPluginDirectory::GetPluginResult(item->GetDynPath(), *item, true);
+          }
           if (item->IsAudio() || item->IsVideo())
             Play(item, pMsg->strParam);
           else
