From a867b73734a04c751c0c8e2b827b97cd45cfafed Mon Sep 17 00:00:00 2001
From: roke <roke@hycu.com>
Date: Fri, 7 Feb 2020 11:10:42 +0100
Subject: [PATCH] fixed checking if proxy and username pass are empty

---
 xbmc/filesystem/CurlFile.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/filesystem/CurlFile.cpp b/xbmc/filesystem/CurlFile.cpp
index edfb94c07d4b..0bbe38dd871c 100644
--- a/xbmc/filesystem/CurlFile.cpp
+++ b/xbmc/filesystem/CurlFile.cpp
@@ -604,7 +604,7 @@ void CCurlFile::SetCommonOptions(CReadState* state, bool failOnError /* = true *
 
     const std::string userpass =
       m_proxyuser + std::string(":") + m_proxypassword;
-    if (!userpass.empty())
+    if (userpass != ":")
       g_curlInterface.easy_setopt(h, CURLOPT_PROXYUSERPWD, userpass.c_str());
   }
   if (m_customrequest.length() > 0)
