From 8456a0ee9155400e29a4b0c89b94640f0fb00995 Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Sat, 6 Jun 2020 10:15:32 +0200
Subject: [PATCH] [Android] Fix stall on stream end

---
 .../DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp          | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp
index 2ca4c2b9a6a4..fe01fb5b26fe 100644
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp
@@ -1044,7 +1044,7 @@ CDVDVideoCodec::VCReturn CDVDVideoCodecAndroidMediaCodec::GetPicture(VideoPictur
   if (!m_opened)
     return VC_NONE;
 
-  if (m_state == MEDIACODEC_STATE_RUNNING &&
+  if (m_state != MEDIACODEC_STATE_FLUSHED &&
       (m_OutputDuration < m_fpsDuration || (m_codecControlFlags & DVD_CODEC_CTRL_DRAIN) != 0))
   {
     m_videobuffer.videoBuffer = pVideoPicture->videoBuffer;
