From 76d6cef3ecbf6463b153ad576e3dda881a03bed8 Mon Sep 17 00:00:00 2001
From: Rechi <Rechi@users.noreply.github.com>
Date: Sat, 14 Sep 2019 11:19:43 +0200
Subject: [PATCH] [cleanup] remove unused variable

---
 xbmc/utils/test/TestVariant.cpp | 1 -
 1 file changed, 1 deletion(-)

diff --git a/xbmc/utils/test/TestVariant.cpp b/xbmc/utils/test/TestVariant.cpp
index 0dfa18292356..fb644fda8653 100644
--- a/xbmc/utils/test/TestVariant.cpp
+++ b/xbmc/utils/test/TestVariant.cpp
@@ -241,7 +241,6 @@ TEST(TestVariant, iterator_array)
     EXPECT_STREQ("string", it->c_str());
   }
 
-  CVariant::const_iterator_array const_it;
   for (auto const_it = a.begin_array(); const_it != a.end_array(); const_it++)
   {
     EXPECT_STREQ("string", const_it->c_str());
