From 3af0abcb11e6f2ee353b1bea7e4cdaf78a91e62f Mon Sep 17 00:00:00 2001
From: DaveTBlake <oak99sky@yahoo.co.uk>
Date: Sun, 15 Mar 2020 13:19:29 +0000
Subject: [PATCH] Make migration from earlier db versions more robust

---
 xbmc/music/MusicDatabase.cpp | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/xbmc/music/MusicDatabase.cpp b/xbmc/music/MusicDatabase.cpp
index 62c829a6952e..4731773df110 100644
--- a/xbmc/music/MusicDatabase.cpp
+++ b/xbmc/music/MusicDatabase.cpp
@@ -7932,12 +7932,13 @@ void CMusicDatabase::UpdateTables(int version)
   if (version < 73)
   {
     // add bBoxedSet to album table
-    m_pDS->exec("ALTER TABLE album ADD bBoxedSet INTEGER \n");
+    m_pDS->exec("ALTER TABLE album ADD bBoxedSet INTEGER NOT NULL DEFAULT 0 \n");
     //  add iDiscTotal to album table
-    m_pDS->exec("ALTER TABLE album ADD iDiscTotal INTEGER \n");
+    m_pDS->exec("ALTER TABLE album ADD iDiscTotal INTEGER NOT NULL DEFAULT 0 \n");
     // populate iDiscTotal from the data already in the song table
     m_pDS->exec("UPDATE album SET iDisctotal = (SELECT COUNT(DISTINCT (iTrack >> 16)) "
-                "FROM song WHERE song.idAlbum = album.idAlbum GROUP BY idAlbum ) ");
+                "FROM song WHERE song.idAlbum = album.idAlbum GROUP BY idAlbum ) "
+                "WHERE EXISTS (SELECT 1 FROM song WHERE song.idAlbum = album.idAlbum)");
     // add strDiscSubtitles to song table
     m_pDS->exec("ALTER TABLE song ADD strDiscSubtitle TEXT \n");
   }
