From f1fdfb2e0f750454759d319c571135220e9b2ce4 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Thu, 14 Mar 2019 01:38:55 +0100
Subject: [PATCH 1/2] choose art > fix thumbnail listing

---
 xbmc/utils/ScraperUrl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/utils/ScraperUrl.cpp b/xbmc/utils/ScraperUrl.cpp
index 9561680d8214..51acc88b27ab 100644
--- a/xbmc/utils/ScraperUrl.cpp
+++ b/xbmc/utils/ScraperUrl.cpp
@@ -377,7 +377,7 @@ void CScraperUrl::GetThumbURLs(std::vector<std::string> &thumbs, const std::stri
 {
   for (std::vector<SUrlEntry>::const_iterator iter = m_url.begin(); iter != m_url.end(); ++iter)
   {
-    if (iter->m_aspect == type || type.empty() || type == "thumb" || iter->m_aspect.empty())
+    if (iter->m_aspect == type || type.empty() || iter->m_aspect.empty())
     {
       if ((iter->m_type == CScraperUrl::URL_TYPE_GENERAL && season == -1)
        || (iter->m_type == CScraperUrl::URL_TYPE_SEASON && iter->m_season == season))

From d85fcfe947b07f38cba59921eb111672d3cbdda4 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Fri, 15 Mar 2019 03:01:11 +0100
Subject: [PATCH 2/2] choose art > don't show thumb were not applicable

---
 xbmc/music/dialogs/GUIDialogMusicInfo.cpp | 16 ----------------
 1 file changed, 16 deletions(-)

diff --git a/xbmc/music/dialogs/GUIDialogMusicInfo.cpp b/xbmc/music/dialogs/GUIDialogMusicInfo.cpp
index 72f52fe65832..dde8506d0faf 100644
--- a/xbmc/music/dialogs/GUIDialogMusicInfo.cpp
+++ b/xbmc/music/dialogs/GUIDialogMusicInfo.cpp
@@ -755,22 +755,6 @@ void CGUIDialogMusicInfo::OnGetArt()
     item->SetLabel(g_localizeStrings.Get(13512));
     items.Add(item);
   }
-  else if (m_item->HasArt("thumb"))
-  {
-    // For missing art of that type add the thumb (when it exists and not a fallback)
-    CGUIListItem::ArtMap::const_iterator i = primeArt.find("thumb");
-    if (i != primeArt.end())
-    {
-      CFileItemPtr item(new CFileItem("thumb://Thumb", false));
-      item->SetArt("thumb", m_item->GetArt("thumb"));
-      if (m_bArtistInfo)
-        item->SetIconImage("DefaultArtistCover.png");
-      else
-        item->SetIconImage("DefaultAlbumCover.png");
-      item->SetLabel(g_localizeStrings.Get(21371));
-      items.Add(item);
-    }
-  }
 
  // Grab the thumbnails of this art type scraped from the web
   std::vector<std::string> remotethumbs;
