From 72601e3a6108aada7fdb91537f75a73119196a17 Mon Sep 17 00:00:00 2001
From: DaveTBlake <oak99sky@yahoo.co.uk>
Date: Tue, 26 May 2020 22:30:43 +0100
Subject: [PATCH] Fix initialising song count later use to reserve JSON results

---
 xbmc/music/MusicDatabase.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/xbmc/music/MusicDatabase.cpp b/xbmc/music/MusicDatabase.cpp
index a553cda927b1..cb1646221942 100644
--- a/xbmc/music/MusicDatabase.cpp
+++ b/xbmc/music/MusicDatabase.cpp
@@ -6573,6 +6573,7 @@ bool CMusicDatabase::GetSongsByWhereJSON(const std::set<std::string>& fields, co
     // (includes xsp limits from filter, but not sort limits)
     total = static_cast<int>(strtol(
         GetSingleValue("SELECT COUNT(1) FROM song " + strSQLExtra, m_pDS).c_str(), nullptr, 10));
+    resultcount = static_cast<size_t>(total);
 
     int iAddedFields = GetOrderFilter(MediaTypeSong, sortDescription, extFilter);
     // Replace songview field names in order by with song, album path table field names
