From 3eed7d28f620def7958814d70c1cbbf559751a5f Mon Sep 17 00:00:00 2001
From: Ryan Rector <rmrector@gmail.com>
Date: Sun, 9 Feb 2020 09:45:01 -0700
Subject: [PATCH] Don't try to generate a thumb for unresolved plugin path

---
 xbmc/video/VideoThumbLoader.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/xbmc/video/VideoThumbLoader.cpp b/xbmc/video/VideoThumbLoader.cpp
index bf6b49696c94..4af2438ba336 100644
--- a/xbmc/video/VideoThumbLoader.cpp
+++ b/xbmc/video/VideoThumbLoader.cpp
@@ -84,6 +84,7 @@ bool CThumbExtractor::DoWork()
   ||  URIUtils::IsPVRRecording(m_item.GetDynPath())
   ||  URIUtils::IsUPnP(m_item.GetPath())
   ||  URIUtils::IsBluray(m_item.GetPath())
+  ||  URIUtils::IsPlugin(m_item.GetDynPath()) // plugin path not fully resolved
   ||  m_item.IsBDFile()
   ||  m_item.IsDVD()
   ||  m_item.IsDiscImage()
