From 085adcee2c751e516ded45a7a5ad92994cea341f Mon Sep 17 00:00:00 2001
From: Lukas Rusak <lorusak@gmail.com>
Date: Mon, 30 Mar 2020 19:48:25 -0700
Subject: [PATCH] CDMAHeapBufferObject: add /dev/dma_heap/linux,cma path

---
 xbmc/utils/DMAHeapBufferObject.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/xbmc/utils/DMAHeapBufferObject.cpp b/xbmc/utils/DMAHeapBufferObject.cpp
index c45db7bb4144..64bcae06f7e4 100644
--- a/xbmc/utils/DMAHeapBufferObject.cpp
+++ b/xbmc/utils/DMAHeapBufferObject.cpp
@@ -22,8 +22,9 @@
 namespace
 {
 
-std::array<const char*, 2> DMA_HEAP_PATHS = {
+std::array<const char*, 3> DMA_HEAP_PATHS = {
     "/dev/dma_heap/reserved",
+    "/dev/dma_heap/linux,cma",
     "/dev/dma_heap/system",
 };
 
