diff --git a/xbmc/guilib/guiinfo/PlayerGUIInfo.cpp b/xbmc/guilib/guiinfo/PlayerGUIInfo.cpp
index 14e0a5cccb..7de52fc5ce 100644
--- a/xbmc/guilib/guiinfo/PlayerGUIInfo.cpp
+++ b/xbmc/guilib/guiinfo/PlayerGUIInfo.cpp
@@ -625,7 +625,8 @@ std::string CPlayerGUIInfo::GetContentRanges(int iInfo) const
     for (const auto& range : ranges)
       values += StringUtils::Format("%.5f,%.5f,", range.first, range.second);
 
-    values.pop_back(); // remove trailing comma
+    if (!values.empty())
+      values.pop_back(); // remove trailing comma
   }
 
   return values;
