From 77ab9b72bd4e62b79e98b911f8a97a2293d646f7 Mon Sep 17 00:00:00 2001
From: James Broadhead <jamesbroadhead@gmail.com>
Date: Mon, 20 May 2019 22:08:20 +0100
Subject: [PATCH] [kodi.sh] pass sigterm to kodi binary

---
 tools/Linux/kodi.sh.in | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/tools/Linux/kodi.sh.in b/tools/Linux/kodi.sh.in
index 4b2b6a747645..73d8c9be814c 100644
--- a/tools/Linux/kodi.sh.in
+++ b/tools/Linux/kodi.sh.in
@@ -192,6 +192,12 @@ print_crash_report()
   echo "Crash report available at $FILE"
 }
 
+propagate_sigterm() {
+  kill -TERM "$CHILD" 2>/dev/null
+}
+
+trap propagate_sigterm TERM
+
 migrate_home
 
 if command_exists gdb; then
@@ -207,7 +213,9 @@ while [ $(( $LOOP )) = "1" ]
 do
   [ -f "${APPORT_CORE}" ] && rm -f "${APPORT_CORE}"
   LOOP=0
-  ${KODI_BINARY} $SAVED_ARGS
+  ${KODI_BINARY} $SAVED_ARGS &
+  CHILD=$!
+  wait "${CHILD}"
   RET=$?
   if [ $(( $RET == 65 )) = "1" ]
   then # User requested to restart app
