From 4ea5997cd766148bad5701a1dd7054cc77a7c699 Mon Sep 17 00:00:00 2001
From: chewitt <christianshewitt@gmail.com>
Date: Mon, 25 May 2020 07:20:42 +0000
Subject: [PATCH] add Samsung exynos to DRMUtils.cpp

---
 xbmc/windowing/gbm/DRMUtils.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/xbmc/windowing/gbm/DRMUtils.cpp b/xbmc/windowing/gbm/DRMUtils.cpp
index 6d198cf520a9..0e96a414e0c1 100644
--- a/xbmc/windowing/gbm/DRMUtils.cpp
+++ b/xbmc/windowing/gbm/DRMUtils.cpp
@@ -594,13 +594,15 @@ bool CDRMUtils::OpenDrm(bool needConnector)
     "radeon",
     "nouveau",
     "vmwgfx",
+    "exynos",
     "msm",
     "imx-drm",
     "rockchip",
     "vc4",
     "virtio_gpu",
-    "sun4i-drm",
+    "mediatek",
     "meson",
+    "sun4i-drm",
     "vboxvideo",
   };
 
