From d39665d92246938e9619f30c5c27e88325a1d2a6 Mon Sep 17 00:00:00 2001
From: MilhouseVH <milhouseVH.github@nmacleod.com>
Date: Mon, 9 Mar 2020 10:27:12 +0000
Subject: [PATCH] SWIG: find also v4

---
 cmake/modules/FindSWIG.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmake/modules/FindSWIG.cmake b/cmake/modules/FindSWIG.cmake
index f9b27c2271e0..fba544baa297 100644
--- a/cmake/modules/FindSWIG.cmake
+++ b/cmake/modules/FindSWIG.cmake
@@ -8,7 +8,7 @@
 # SWIG_FOUND - system has SWIG
 # SWIG_EXECUTABLE - the SWIG executable
 
-find_program(SWIG_EXECUTABLE NAMES swig3.0 swig2.0 swig
+find_program(SWIG_EXECUTABLE NAMES swig4.0 swig3.0 swig2.0 swig
                              PATH_SUFFIXES swig)
 if(SWIG_EXECUTABLE)
   execute_process(COMMAND ${SWIG_EXECUTABLE} -swiglib
