From 27c4f4a824a72c5a1ea3ae5aa668b07613541b18 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Wed, 22 Jan 2020 02:49:06 +0100
Subject: [PATCH] [videos] allow 0.0 rating

---
 xbmc/guilib/guiinfo/VideoGUIInfo.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/xbmc/guilib/guiinfo/VideoGUIInfo.cpp b/xbmc/guilib/guiinfo/VideoGUIInfo.cpp
index 450b1ff68c6b..9806208d0c99 100644
--- a/xbmc/guilib/guiinfo/VideoGUIInfo.cpp
+++ b/xbmc/guilib/guiinfo/VideoGUIInfo.cpp
@@ -154,14 +154,16 @@ bool CVideoGUIInfo::GetLabel(std::string& value, const CFileItem *item, int cont
       case LISTITEM_RATING_AND_VOTES:
       {
         CRating rating = tag->GetRating(info.GetData3());
-        if (rating.rating > 0.f)
+        if (rating.rating >= 0.f)
         {
-          if (rating.votes == 0)
+          if (rating.rating > 0.f && rating.votes == 0)
             value = StringUtils::FormatNumber(rating.rating);
-          else
+          else if (rating.votes > 0)
             value = StringUtils::Format(g_localizeStrings.Get(20350).c_str(),
                                         StringUtils::FormatNumber(rating.rating).c_str(),
                                         StringUtils::FormatNumber(rating.votes).c_str());
+          else
+            break;
           return true;
         }
         break;
