From 1f10c7571a22f194fe0fa2c5a1ccf24619576a0c Mon Sep 17 00:00:00 2001
From: phunkyfish <phunkyfish@gmail.com>
Date: Tue, 19 Nov 2019 19:28:17 +0000
Subject: [PATCH 1/2] [pvr] refactoring fix whitespace

---
 xbmc/utils/URIUtils.cpp | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/xbmc/utils/URIUtils.cpp b/xbmc/utils/URIUtils.cpp
index f0f84e45173f..ebdae3122df7 100644
--- a/xbmc/utils/URIUtils.cpp
+++ b/xbmc/utils/URIUtils.cpp
@@ -600,16 +600,16 @@ bool URIUtils::IsRemote(const std::string& strFile)
 
   if (IsSourcesPath(strFile))
     return false;
-  
+
   if (IsVideoDb(strFile) || IsMusicDb(strFile))
     return false;
-  
+
   if (IsLibraryFolder(strFile))
     return false;
-  
+
   if (IsPlugin(strFile))
     return false;
-  
+
   if (IsAndroidApp(strFile))
     return false;
 

From 4ecbadf0c7f1e1b086a87922c6a6ac959b1ec9f4 Mon Sep 17 00:00:00 2001
From: phunkyfish <phunkyfish@gmail.com>
Date: Tue, 19 Nov 2019 19:28:34 +0000
Subject: [PATCH 2/2] [pvr] refactor recording file or folder check

---
 xbmc/utils/URIUtils.cpp     | 5 +++++
 xbmc/utils/URIUtils.h       | 1 +
 xbmc/video/ContextMenus.cpp | 4 ++--
 3 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/xbmc/utils/URIUtils.cpp b/xbmc/utils/URIUtils.cpp
index ebdae3122df7..c564e0aade37 100644
--- a/xbmc/utils/URIUtils.cpp
+++ b/xbmc/utils/URIUtils.cpp
@@ -1063,6 +1063,11 @@ bool URIUtils::IsPVRRecording(const std::string& strFile)
          StringUtils::StartsWith(strFile, "pvr://recordings");
 }
 
+bool URIUtils::IsPVRRecordingFileOrFolder(const std::string& strFile)
+{
+  return StringUtils::StartsWith(strFile, "pvr://recordings");
+}
+
 bool URIUtils::IsMusicDb(const std::string& strFile)
 {
   return IsProtocol(strFile, "musicdb");
diff --git a/xbmc/utils/URIUtils.h b/xbmc/utils/URIUtils.h
index f10e4a204283..14f962a03da7 100644
--- a/xbmc/utils/URIUtils.h
+++ b/xbmc/utils/URIUtils.h
@@ -129,6 +129,7 @@ class URIUtils
   static bool IsISO9660(const std::string& strFile);
   static bool IsLiveTV(const std::string& strFile);
   static bool IsPVRRecording(const std::string& strFile);
+  static bool IsPVRRecordingFileOrFolder(const std::string& strFile);
   static bool IsMultiPath(const std::string& strPath);
   static bool IsMusicDb(const std::string& strFile);
   static bool IsNfs(const std::string& strFile);
diff --git a/xbmc/video/ContextMenus.cpp b/xbmc/video/ContextMenus.cpp
index da3198718e7f..4e02c38ff58e 100644
--- a/xbmc/video/ContextMenus.cpp
+++ b/xbmc/video/ContextMenus.cpp
@@ -65,7 +65,7 @@ bool CMarkWatched::IsVisible(const CFileItem& item) const
     else if (item.GetProperty("IsVideoFolder").asBoolean())
       return true;
     else
-      return StringUtils::StartsWith(item.GetPath(), "pvr://recordings/");
+      return URIUtils::IsPVRRecordingFileOrFolder(item.GetPath());
   }
   else if (!item.HasVideoInfoTag())
     return false;
@@ -91,7 +91,7 @@ bool CMarkUnWatched::IsVisible(const CFileItem& item) const
     else if (item.GetProperty("IsVideoFolder").asBoolean())
       return true;
     else
-      return StringUtils::StartsWith(item.GetPath(), "pvr://recordings/");
+      return URIUtils::IsPVRRecordingFileOrFolder(item.GetPath());
   }
   else if (!item.HasVideoInfoTag())
     return false;
