From ed4a38c127a0f7b8ddc0c7df43e87b1bfc69b4b3 Mon Sep 17 00:00:00 2001
From: ronie <ronie@kodi.tv>
Date: Sun, 19 Jan 2020 14:37:31 +0100
Subject: [PATCH] Dialog.Close(all) should only close modal dialogs

---
 xbmc/guilib/GUIWindowManager.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/xbmc/guilib/GUIWindowManager.cpp b/xbmc/guilib/GUIWindowManager.cpp
index e10d41d0a034..4b2644197b98 100644
--- a/xbmc/guilib/GUIWindowManager.cpp
+++ b/xbmc/guilib/GUIWindowManager.cpp
@@ -851,7 +851,8 @@ void CGUIWindowManager::CloseDialogs(bool forceClose) const
   auto activeDialogs = m_activeDialogs;
   for (const auto& window : activeDialogs)
   {
-    window->Close(forceClose);
+    if (window->IsModalDialog())
+      window->Close(forceClose);
   }
 }
 
