From 709afc29e490236ef0361cfa4613f7bab071a0dc Mon Sep 17 00:00:00 2001
From: Greg McCarthy <greg@gjmccarthy.co.uk>
Date: Wed, 16 Oct 2019 12:23:52 +0100
Subject: [PATCH] Remove python patch

---
 xbmc/interfaces/generic/LanguageInvokerThread.cpp | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/xbmc/interfaces/generic/LanguageInvokerThread.cpp b/xbmc/interfaces/generic/LanguageInvokerThread.cpp
index e9ff8d754c..20480267dc 100644
--- a/xbmc/interfaces/generic/LanguageInvokerThread.cpp
+++ b/xbmc/interfaces/generic/LanguageInvokerThread.cpp
@@ -52,14 +52,8 @@ bool CLanguageInvokerThread::execute(const std::string &script, const std::vecto
     m_condition.notify_one();
   }
   else
-  {
     Create();
 
-    #ifdef TARGET_RASPBERRY_PI
-    /* low prio */
-    SetPriority(GetPriority()-1);
-    #endif
-  }
   //Todo wait until running
 
   return true;
-- 
2.16.4

